homeassistant/components/switchbot/config_flow.py

```diff
@@ -84,36 +71,58 @@ def async_get_options_flow(
         """Get the options flow for this handler."""
         return SwitchbotOptionsFlowHandler(config_entry)
 
+    def __init__(self):
+        """Initialize the config flow."""
+        self._discovered_devices = {}
+
     async def async_step_user(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle a flow initiated by the user."""
 
-        errors = {}
+        errors: dict[str, str] = {}
 
         if user_input is not None:
-            user_input[CONF_MAC] = user_input[CONF_MAC].replace("-", ":").lower()
+            await self.async_set_unique_id(user_input[CONF_MAC].replace(":", ""))
+            self._abort_if_unique_id_configured()
+
+            user_input[CONF_SENSOR_TYPE] = SUPPORTED_MODEL_TYPES[
+                self._discovered_devices[self.unique_id]["modelName"]
+            ]
 
-            # abort if already configured.
-            for item in self._async_current_entries():
-                if item.data.get(CONF_MAC) == user_input[CONF_MAC]:
-                    return self.async_abort(reason="already_configured_device")
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
 
-            try:
-                return await self._validate_mac(user_input)
+        try:
+            self._discovered_devices = await self._get_switchbots()
 
-            except NotConnectedError:
-                errors["base"] = "cannot_connect"
+        except NotConnectedError:
+            return self.async_abort(reason="cannot_connect")
+
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            return self.async_abort(reason="unknown")
+
+        # Get devices already configured.
+        configured_devices = {
+            item.data.get(CONF_MAC) for item in self._async_current_entries()
+        }
+
+        # Get supported devices not yet configured.
+        unconfigured_devices = {
+            device["mac_address"]
+            for device in self._discovered_devices.values()
+            if device.get("modelName") in SUPPORTED_MODEL_TYPES
+            if device.get("mac_address") not in configured_devices
```

Comment:
```suggestion
            if device["mac_address"] not in configured_devices
```

Reply:
These can't be missing

https://github.com/Danielhiversen/pySwitchbot/blob/master/switchbot/__init__.py#L371